### PR TITLE
Fixes #19715 - [Check] WARN if foreman-proxy is self-registered

### PR DESCRIPTION
**Pull request by swapab**
Raise WARNING for smart-proxy if self-registered.
REF: https://github.com/theforeman/foreman_maintain/issues/23#issuecomment-300745839

### CODE REVIEW

**Comment by swapab** [comment]
I have piggybacked `katello` feature. Which would confine disk_io check to katello's presence.

File: definitions/features/katello.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Features::Katello < ForemanMaintain::Feature
+  metadata do
+    label :katello
+  end
+
+  def dirs
```
Comment (iNecas) [inline definitions/features/katello.rb]: The `dirs` is too generic. Can you thing of something more specific? `data_dirs` perhaps?

File: lib/foreman_maintain/concerns/system_helpers.rb [hunk 2]
```diff
@@ -89,9 +50,60 @@ def execute(command, options = {})
...
+
+      def install_packages(packages, options = {})
+        options.validate_options!(:assumeyes)
+        yum_options = []
+        yum_options << '-y' if options[:assumeyes]
+        execute!("yum #{yum_options.join(' ')} install #{packages.join(' ')}", :interactive => true)
+      end
+
+      def server?
+        find_package('foreman')
```
Comment (iNecas) [inline lib/foreman_maintain/concerns/system_helpers.rb]: shouldn't this be `foreman` and `foreman-proxy` features?

File: definitions/features/katello.rb (hunk 1 above)
Comment (swapab) [inline definitions/features/katello.rb]: Yes makes sense. Done :+1:

File: lib/foreman_maintain/concerns/system_helpers.rb (hunk 2 above)
Comment (swapab) [inline lib/foreman_maintain/concerns/system_helpers.rb]: Absolutely!
Added new features. `foreman-server` and `foreman-proxy` to be more specific and `label :foreman` was already taken.

File: definitions/checks/system_registration.rb [hunk 3]
```diff
@@ -10,15 +10,23 @@ class Checks::SystemRegistration < ForemanMaintain::Check
...
 
-  def system_is_self_registerd?
-    rhsm_hostname.casecmp(hostname).zero?
+  def system_self_registerd?
+    @result ||= rhsm_hostname.casecmp(hostname).zero?
+  end
+
+  def warn?
+    if feature(:foreman_server).present?
+      !system_self_registerd?
```
Comment (iNecas) [inline definitions/checks/system_registration.rb]: If I read it properly, we would warn if run on main server and the system was not self-registered. Despite the fact, that the error message would be `System is self registered`, is this what we actually want?

Comment (swapab) [inline definitions/checks/system_registration.rb]: Per Lukas's comment, a warning should be raised if the system is a **smart-proxy and self-registered**.
https://github.com/theforeman/foreman_maintain/issues/23#issuecomment-300745839

Comment (swapab) [inline definitions/checks/system_registration.rb]: I have updated the check `Checks::SystemRegistration` to check self-registration if **only smart-proxy present** and foreman-server is absent

**Comment by swapab** [comment]
@iNecas I have updated the PR with following changes:
1. Confine `feature(:foreman_server)` with package presence.
2. Confine `check(:verify_self_registration)` with condition `Checks::SystemRegistration#system_self_registered?`

File: definitions/features/foreman_proxy.rb [hunk 4]
```diff
@@ -5,6 +5,10 @@ class Features::ForemanProxy < ForemanMaintain::Feature
 
   attr_reader :dhcpd_conf_file, :cert_path
 
+  def present?
+    !feature(:foreman_server).present? && find_package('foreman-proxy')
```
Comment (iNecas) [inline definitions/features/foreman_proxy.rb]: You can have foreman_proxy feature present also at the same machine as the foreman. This condition probably belongs to the confine block of the self-registered check itself

File: definitions/checks/system_registration.rb [hunk 5]
```diff
@@ -5,27 +5,32 @@ class Checks::SystemRegistration < ForemanMaintain::Check
...
+    def rhsm_hostname_eql_hostname?
+      @result ||= rhsm_hostname.casecmp(hostname).zero?
+    end
 
-  def rhsm_conf_file
-    '/etc/rhsm/rhsm.conf'
+    def system_self_registered?
+      file_exists?('/etc/rhsm/rhsm.conf') &&
+        !feature(:foreman_server) &&
+        feature(:foreman_proxy).present? &&
```
Comment (iNecas) [inline definitions/checks/system_registration.rb]: Similarly as the `:foreman_server` also the `:foreman_proxy` and `:katello` should rather use confine blocks than the `present?` method, to match with the other features

Comment (iNecas) [reply definitions/checks/system_registration.rb]: Also, this condition should go to the confine block of this check

File: definitions/features/foreman_proxy.rb (hunk 4 above)
Comment (swapab) [reply definitions/features/foreman_proxy.rb]: Ah, right. That's correct `present?` is redundant here.

File: definitions/checks/system_registration.rb (hunk 5 above)
Comment (swapab) [reply definitions/checks/system_registration.rb]: You mean the whole condition of `system_self_registered` should belong to confine block? OR just
```
confine do
  file_exists?('/etc/rhsm/rhsm.conf') &&
    !feature(:foreman_server) &&
    feature(:foreman_proxy)
end
```

**Comment by swapab** [comment]
I have updated the checks accordingly but, the tests have become pretty slow now.
I think it's because of the package presence checks in the respective confine blocks.

```
Finished in 55.495118s, 0.5946 runs/s, 1.3154 assertions/s.

33 runs, 73 assertions, 0 failures, 0 errors, 0 skips
```
It used to execute in ~20 seconds before. Need to find the slow tests and stub confine blocks for them.

File: lib/foreman_maintain/concerns/system_helpers.rb [hunk 6]
```diff
@@ -26,16 +26,66 @@ def build
...
       end
 
       def file_exists?(filename)
         File.exist?(filename)
       end
 
-      def version(value)
-        Version.new(value)
+      def find_package(name)
+        execute(%(rpm -qa '#{name}'))
```
Comment (iNecas) [inline lib/foreman_maintain/concerns/system_helpers.rb]: the `qa` variant is too slow. `rpm -q ` should be used instead. Also make sure it returns nil/false if the pacakge is not found: currently it returns '' instead: therefore the feature looks like it's present even though it isn't

**Comment by swapab** [comment]
Requested changes were done. Failing tests fixed.

File: lib/foreman_maintain/concerns/system_helpers.rb (hunk 6 above)
Comment (swapab) [reply lib/foreman_maintain/concerns/system_helpers.rb]: Tests running in `< 6 seconds` now.
Thanks! @iNecas

**Comment by iNecas** [comment]
Tested and works well now. Thanks @swapab

**Pull request merged**